core/lib/zksync_core/src/state_keeper/metrics.rs

```diff
@@ -0,0 +1,378 @@
+    #[metrics(buckets = Buckets::LATENCIES)]
+    pub load_previous_miniblock_header: Histogram<Duration>,
+    /// Time spent by the state keeper on transaction execution.
+    #[metrics(buckets = Buckets::LATENCIES)]
+    pub tx_execution_time: Family<TxExecutionStage, Histogram<Duration>>,
+    /// Number of times gas price was reported as too high.
+    pub gas_price_too_high: Counter,
+}
+
+#[vise::register]
+pub(crate) static KEEPER_METRICS: vise::Global<StateKeeperMetrics> = vise::Global::new();
+
+/// State keeper-related gauges exposed via a collector.
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "server_state_keeper")]
+pub(super) struct StateKeeperGauges {
+    /// Current number of L1 transactions in the mempool.
+    mempool_l1_size: Gauge<usize>,
+    /// Current number of L2 transactions in the mempool.
+    mempool_l2_size: Gauge<u64>,
+    /// Current size of the L2 priority queue.
+    l2_priority_queue_size: Gauge<usize>,
+}
+
+impl StateKeeperGauges {
+    pub(super) fn register(pool_ref: Weak<Mutex<MempoolStore>>) {
+        #[vise::register]
+        static COLLECTOR: vise::Collector<Option<StateKeeperGauges>> = vise::Collector::new();
+
+        let res = COLLECTOR.before_scrape(move || {
+            pool_ref.upgrade().map(|pool| {
+                let stats = pool.lock().expect("failed to acquire mempool lock").stats();
+                drop(pool); // Don't prevent the pool to be dropped
+
+                let gauges = StateKeeperGauges::default();
+                gauges.mempool_l1_size.set(stats.l1_transaction_count);
+                gauges.mempool_l2_size.set(stats.l2_transaction_count);
+                gauges
+                    .l2_priority_queue_size
+                    .set(stats.l2_priority_queue_size);
+                gauges
+            })
+        });
+        if res.is_err() {
+            tracing::warn!(
+                "Mempool registered for metrics multiple times; this is a logical error"
+            );
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue)]
+#[metrics(rename_all = "snake_case")]
+enum SealResolutionLabel {
+    NoSeal,
+    IncludeAndSeal,
+    ExcludeAndSeal,
+    Unexecutable,
+}
+
+impl From<&SealResolution> for SealResolutionLabel {
+    fn from(resolution: &SealResolution) -> Self {
+        match resolution {
+            SealResolution::NoSeal => Self::NoSeal,
+            SealResolution::IncludeAndSeal => Self::IncludeAndSeal,
+            SealResolution::ExcludeAndSeal => Self::ExcludeAndSeal,
+            SealResolution::Unexecutable(_) => Self::Unexecutable,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelSet)]
+struct TxAggregationLabels {
+    criterion: &'static str,
+    seal_resolution: Option<SealResolutionLabel>,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "server_tx_aggregation")]
+pub(super) struct TxAggregationMetrics {
```

Comment:
This should probably be a part of `StateKeeperMetrics`; it's only a separate type because it has a differing prefix.